include/merlin_hashtable.cuh

```diff
@@ -424,9 +420,10 @@ class HashTable {
     int* src_offset;
     bool* founds;
 
-    std::shared_lock<std::shared_timed_mutex> lock(mutex_, std::defer_lock);
-    if (!reach_max_capacity_) {
-      lock.lock();
+    // Block exclusive access, unless table structure cannot change anymore.
+    std::shared_lock<std::shared_timed_mutex> lock(table_mutex_);
```

Comment:
I think the lock should be always called, it has overhead as my test. So the `defer_lock` will be helpful in performance.

Reply:
The problem is that If you do not lock you risk a race condition in very fast multi-threading situations.

Reply:
IMHO: You would have to make the `bool` volatile to ensure it always works.

Reply:
But if you make it `volatile`, you probably get the same performance penalty as if you had the lock. Likely, that is about the way the lock is implemented.

Reply:
https://stackoverflow.com/questions/3652056/how-efficient-is-locking-and-unlocked-mutex-what-is-the-cost-of-a-mutex

Even on much slower CPUs, it takes about 100ns to lock/unlock a mutex. I doubt you could measure it.

Reply:
Quote: ```Using std::mutex takes about 74 machine cycles```

Reply:
I fixed it according to your idea. See inline comments to understand what I did, and where I think there are still isues.